optuna/integration/pytorch_lightning.py

```diff
@@ -52,7 +69,41 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
             warnings.warn(message)
             return
 
-        self._trial.report(current_score, step=epoch)
-        if self._trial.should_prune():
+        if not self.is_ddp_backend:
+            self._trial.report(current_score, step=epoch)
+            if self._trial.should_prune():
+                message = "Trial was pruned at epoch {}.".format(epoch)
+                raise optuna.TrialPruned(message)
+        else:
+            try:
+                self._trial.report(current_score, step=epoch)
+            except sqlalchemy.exc.IntegrityError:
+                pass
+
+            should_stop = self._trial.should_prune()
+            # stop every ddp process if any world process decides to stop
+            should_stop = trainer.training_type_plugin.reduce_boolean_decision(should_stop)
+            if should_stop:
+                trainer.should_stop = True
+                if trainer.is_global_zero:
+                    self._trial.set_system_attr("pruned", True)
+                    self._trial.set_system_attr("epoch", epoch)
```

Comment:
How about this?


```suggestion
        if trainer.is_global_zero:
            self._trial.report(current_score, step=epoch)
            should_stop = self._trial.should_prune()
        should_stop = trainer.training_type_plugin.reduce_boolean_decision(should_stop)

        if not should_stop:
            return
            
        if not self.is_ddp_backend:
            message = "Trial was pruned at epoch {}.".format(epoch)
            raise optuna.TrialPruned(message)
        else:            
            # stop every ddp process if global rank 0 process decides to stop
            trainer.should_stop = True
            if trainer.is_global_zero:
                self._trial.set_system_attr("pruned", True)
                self._trial.set_system_attr("epoch", epoch)
```

Reply:
I think the usage of reduce_boolean_decision is wrong. I actually tried the above code, but it ended with an error.

Reply:
Could you explain why or some example code? The logic is similar to early-stopping in pl.
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/early_stopping.py#L209

Reply:
Indeed, you are right. We might need to use `trainer.training_type_plugin.broadcast` instead of it.

Reply:
I've updated my suggested code. Could you take a look?

Reply:
Thank you.
I updated the source code.